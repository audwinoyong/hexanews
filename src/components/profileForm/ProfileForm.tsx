/**
 * The ProfileForm Component for editting user in account container
 */

import { connect } from 'react-redux'
import { Dispatch } from 'redux'
import { translate } from 'react-i18next'
import { Typography } from '@material-ui/core'
import { ValidatorForm, TextValidator } from 'react-material-ui-form-validator'
import * as React from 'react'
import Button from '@material-ui/core/Button'
import CircularProgress from '@material-ui/core/CircularProgress'
import TextField from '@material-ui/core/TextField'

import { User } from '../../domain/model/User'
import * as actions from './actions'
import i18n from '../../i18n'
import selector, { StateProps } from './selector'
import styles from './styles'
import SuccessSnackbar from '../successSnackbar/SuccessSnackbar'

/**
 * All props required by the components
 */
export type Props = {
  dispatch: Dispatch<any>,
  isEditingUser: boolean,
  user: User,
} & StateProps

/**
 * All state required by the components
 */
export interface ComponentState {
  description: string
  email: string
  isSnackbarOpen: boolean
  name: string
}

export class ProfileForm extends React.Component<Props, ComponentState> {

  /**
   * Main constructor for Profile form and initialize state
   * @param props props required in the components
   */
  constructor(props: Props) {
    super(props)
    const { user } = this.props
    this.state = {
      description: user.description ? user.description : '',
      email: user.email,
      isSnackbarOpen: false,
      name: user.name,
    }
  }

  /**
   * Function that handle onChange text in name textfield
   * @param event contain the name value from textField
   */
  handleNameChange = (event: any) => {
    this.setState({ name: event.target.value })
  }

  /**
   * Function that handle onChange text in description textfield
   * @param event contain the description value from textField
   */
  handleDescriptionChange = (event: any) => {
    this.setState({ description: event.target.value })
  }

  /**
   * Function that disable saveProfile button when name is empty
   */
  disableSaveProfileButton = () => {
    const { name } = this.state
    return (name === '')
  }

  /**
   * Function that call editUser action to save the user profile
   */
  onSaveProfile = async () => {
    const { dispatch } = this.props
    const { email, name, description } = this.state
    const _id = localStorage.getItem('id') || ''

    const edittedUser = { _id, email, name, description }
    await dispatch(actions.editUser(edittedUser))

    const { editUserError } = this.props

    if (editUserError === '') {
      this.setState({ isSnackbarOpen: true })
    }
  }

  /**
   * Function to close snackbar after editUser success
   */
  handleClose = () => {
    this.setState({ isSnackbarOpen: false })
  }

  /**
   * Render Profile form component
   */
  public render() {
    const { isEditingUser, editUserError } = this.props
    const { email, name, description, isSnackbarOpen } = this.state
    return (
      <div style={styles.sectionContainer}>
        <ValidatorForm
          ref='saveProFileForm'
          onSubmit={() => this.onSaveProfile()}
          instantValidate={false}
        >
          <Typography
            variant='headline'
            component='h2'
            color='textPrimary'
            gutterBottom
          >
            {i18n.t('profileForm.editProfile')}
          </Typography>
          <TextField
            id='email'
            label={i18n.t('profileForm.email')}
            value={email}
            helperText=' '
            disabled
            style={styles.textField}
            margin='dense'
          />
          <TextValidator
            label={i18n.t('profileForm.name')}
            onChange={this.handleNameChange}
            name='name'
            style={styles.textField}
            margin='dense'
            value={name}
            helperText=' '
            validators={['required', 'minStringLength:3', 'maxStringLength:50', 'matchRegexp:^[a-zA-Z\\s]+$']}
            errorMessages={[i18n.t('profileForm.enterName'),
              i18n.t('profileForm.minName'),
              i18n.t('profileForm.maxName'),
              i18n.t('profileForm.alphabeticName')]}
          />
          <TextValidator
            label={i18n.t('profileForm.description')}
            onChange={this.handleDescriptionChange}
            name='description'
            style={styles.textField}
            margin='dense'
            value={description}
            helperText=' '
            validators={['maxStringLength:100']}
            errorMessages={[i18n.t('profileForm.maxDescription')]}
          />
          <Typography
            variant='body1'
            component='h2'
            color='textPrimary'
            style={styles.errorChangeLabel}
          >
            {editUserError}
          </Typography>
          <Button
            variant='outlined'
            size='small'
            component='button'
            style={styles.button}
            disabled={this.disableSaveProfileButton()}
            type='submit'
          >
            {isEditingUser ? <CircularProgress size={22} /> : i18n.t('profileForm.saveButton')}
          </Button>
          <SuccessSnackbar
            isSnackbarOpen={isSnackbarOpen}
            message={i18n.t('profileForm.editUserSuccess')}
            handleClose={this.handleClose}
          />
        </ValidatorForm>
      </div>
    )
  }
}

const ConnectedProfileForm = connect(selector)(ProfileForm)

export default translate('translations')(ConnectedProfileForm)