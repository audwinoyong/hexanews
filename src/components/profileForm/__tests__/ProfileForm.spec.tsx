/**
 * Unit tests for checking the behavior of profileForm components
 */

import * as React from 'react'
import * as renderer from 'react-test-renderer'
import * as Enzyme from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'
import { shallow } from 'enzyme'

import { ProfileForm, Props, ComponentState } from '../ProfileForm'

jest.mock('react-quill')
Enzyme.configure({ adapter: new Adapter() })

const mockUser = {
  _id: '1',
  createdAt: 'created',
  description: 'desc',
  email: 'user@email.com',
  name: 'user'
}

describe('Profile Form', () => {
  it('renders Profile Form', () => {
    const component = renderer.create(
      <ProfileForm
        user={mockUser}
        dispatch={jest.fn()}
        isEditingUser={false}
        editUserError=''
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders Profile Form with isEditingUser as true', () => {
    const component = renderer.create(
      <ProfileForm
        user={mockUser}
        dispatch={jest.fn()}
        isEditingUser={true}
        editUserError=''
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('handle handleNameChange', () => {
    const component = (
      <ProfileForm
        user={mockUser}
        dispatch={jest.fn()}
        isEditingUser={true}
        editUserError=''
      />
    )

    const layout = shallow<Props, ComponentState>(component)
    const instance: ProfileForm = layout.instance() as ProfileForm

    const name = layout.state().name
    expect(name).toBe('user')

    const event = {
      target: {
        value: 'user2'
      }
    }

    instance.handleNameChange(event)
    layout.update()

    const updatedName = layout.state().name
    expect(updatedName).toBe('user2')
  })

  it('handle onChangeDescription', () => {
    const component = (
      <ProfileForm
        user={mockUser}
        dispatch={jest.fn()}
        isEditingUser={true}
        editUserError=''
      />
    )

    const layout = shallow<Props, ComponentState>(component)
    const instance: ProfileForm = layout.instance() as ProfileForm

    const description = layout.state().description
    expect(description).toBe('desc')

    const event = {
      target: {
        value: 'desc2'
      }
    }

    instance.handleDescriptionChange(event)
    layout.update()

    const updatedDescription = layout.state().description
    expect(updatedDescription).toBe('desc2')
  })

  it('handle onSaveProfile action correctly', () => {
    const dispatch = jest.fn()
    const component = (
      <ProfileForm
        user={mockUser}
        dispatch={dispatch}
        isEditingUser={true}
        editUserError=''
      />
    )

    const layout = shallow(component)
    const instance: ProfileForm = layout.instance() as ProfileForm

    instance.onSaveProfile()
    layout.update()

    expect(dispatch).toHaveBeenCalled()
  })
})