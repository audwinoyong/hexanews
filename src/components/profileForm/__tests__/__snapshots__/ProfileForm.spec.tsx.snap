// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Form renders Profile Form 1`] = `
<div
  style={
    Object {
      "marginTop": 24,
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <h2
      className="MuiTypography-root-1 MuiTypography-headline-6 MuiTypography-colorTextPrimary-23 MuiTypography-gutterBottom-18"
    >
      Edit Profile
    </h2>
    <div
      aria-describedby="email-helper-text"
      className="MuiFormControl-root-26 MuiFormControl-marginDense-28"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root-35 MuiFormLabel-filled-39 MuiInputLabel-root-30 MuiInputLabel-formControl-31 MuiInputLabel-animated-34 MuiInputLabel-shrink-33 MuiInputLabel-marginDense-32"
        data-shrink={true}
        htmlFor="email"
      >
        Email
      </label>
      <div
        className="MuiInput-root-42 MuiInput-disabled-45 MuiInput-formControl-43 MuiInput-underline-46"
      >
        <input
          aria-invalid={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-50 MuiInput-disabled-45 MuiInput-inputMarginDense-51"
          defaultValue={undefined}
          disabled={true}
          id="email"
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={false}
          rows={undefined}
          type="text"
          value="user@email.com"
        />
      </div>
      <p
        className="MuiFormHelperText-root-55 MuiFormHelperText-filled-60 MuiFormHelperText-marginDense-58"
        id="email-helper-text"
      >
         
      </p>
    </div>
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-26 MuiFormControl-marginDense-28"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root-35 MuiFormLabel-filled-39 MuiInputLabel-root-30 MuiInputLabel-formControl-31 MuiInputLabel-animated-34 MuiInputLabel-shrink-33 MuiInputLabel-marginDense-32"
        data-shrink={true}
        htmlFor={undefined}
      >
        Name
      </label>
      <div
        className="MuiInput-root-42 MuiInput-formControl-43 MuiInput-underline-46"
      >
        <input
          aria-invalid={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-50 MuiInput-inputMarginDense-51"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={false}
          rows={undefined}
          type="text"
          value="user"
        />
      </div>
      <p
        className="MuiFormHelperText-root-55 MuiFormHelperText-filled-60 MuiFormHelperText-marginDense-58"
        id={undefined}
      >
         
      </p>
    </div>
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-26 MuiFormControl-marginDense-28"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root-35 MuiFormLabel-filled-39 MuiInputLabel-root-30 MuiInputLabel-formControl-31 MuiInputLabel-animated-34 MuiInputLabel-shrink-33 MuiInputLabel-marginDense-32"
        data-shrink={true}
        htmlFor={undefined}
      >
        Description
      </label>
      <div
        className="MuiInput-root-42 MuiInput-formControl-43 MuiInput-underline-46"
      >
        <input
          aria-invalid={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-50 MuiInput-inputMarginDense-51"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name="description"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={false}
          rows={undefined}
          type="text"
          value="desc"
        />
      </div>
      <p
        className="MuiFormHelperText-root-55 MuiFormHelperText-filled-60 MuiFormHelperText-marginDense-58"
        id={undefined}
      >
         
      </p>
    </div>
    <h2
      className="MuiTypography-root-1 MuiTypography-body1-10 MuiTypography-colorTextPrimary-23"
      style={
        Object {
          "color": "#E53935",
        }
      }
    >
      
    </h2>
    <button
      className="MuiButtonBase-root-88 MuiButton-root-62 MuiButton-text-64 MuiButton-outlined-70 MuiButton-sizeSmall-85"
      disabled={false}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "marginTop": 10,
        }
      }
      tabIndex="0"
      type="submit"
    >
      <span
        className="MuiButton-label-63"
      >
        Save Profile
      </span>
      <span
        className="MuiTouchRipple-root-91"
      />
    </button>
  </form>
</div>
`;

exports[`Profile Form renders Profile Form with isEditingUser as true 1`] = `
<div
  style={
    Object {
      "marginTop": 24,
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <h2
      className="MuiTypography-root-1 MuiTypography-headline-6 MuiTypography-colorTextPrimary-23 MuiTypography-gutterBottom-18"
    >
      Edit Profile
    </h2>
    <div
      aria-describedby="email-helper-text"
      className="MuiFormControl-root-26 MuiFormControl-marginDense-28"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root-35 MuiFormLabel-filled-39 MuiInputLabel-root-30 MuiInputLabel-formControl-31 MuiInputLabel-animated-34 MuiInputLabel-shrink-33 MuiInputLabel-marginDense-32"
        data-shrink={true}
        htmlFor="email"
      >
        Email
      </label>
      <div
        className="MuiInput-root-42 MuiInput-disabled-45 MuiInput-formControl-43 MuiInput-underline-46"
      >
        <input
          aria-invalid={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-50 MuiInput-disabled-45 MuiInput-inputMarginDense-51"
          defaultValue={undefined}
          disabled={true}
          id="email"
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={false}
          rows={undefined}
          type="text"
          value="user@email.com"
        />
      </div>
      <p
        className="MuiFormHelperText-root-55 MuiFormHelperText-filled-60 MuiFormHelperText-marginDense-58"
        id="email-helper-text"
      >
         
      </p>
    </div>
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-26 MuiFormControl-marginDense-28"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root-35 MuiFormLabel-filled-39 MuiInputLabel-root-30 MuiInputLabel-formControl-31 MuiInputLabel-animated-34 MuiInputLabel-shrink-33 MuiInputLabel-marginDense-32"
        data-shrink={true}
        htmlFor={undefined}
      >
        Name
      </label>
      <div
        className="MuiInput-root-42 MuiInput-formControl-43 MuiInput-underline-46"
      >
        <input
          aria-invalid={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-50 MuiInput-inputMarginDense-51"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={false}
          rows={undefined}
          type="text"
          value="user"
        />
      </div>
      <p
        className="MuiFormHelperText-root-55 MuiFormHelperText-filled-60 MuiFormHelperText-marginDense-58"
        id={undefined}
      >
         
      </p>
    </div>
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-26 MuiFormControl-marginDense-28"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root-35 MuiFormLabel-filled-39 MuiInputLabel-root-30 MuiInputLabel-formControl-31 MuiInputLabel-animated-34 MuiInputLabel-shrink-33 MuiInputLabel-marginDense-32"
        data-shrink={true}
        htmlFor={undefined}
      >
        Description
      </label>
      <div
        className="MuiInput-root-42 MuiInput-formControl-43 MuiInput-underline-46"
      >
        <input
          aria-invalid={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-50 MuiInput-inputMarginDense-51"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name="description"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={false}
          rows={undefined}
          type="text"
          value="desc"
        />
      </div>
      <p
        className="MuiFormHelperText-root-55 MuiFormHelperText-filled-60 MuiFormHelperText-marginDense-58"
        id={undefined}
      >
         
      </p>
    </div>
    <h2
      className="MuiTypography-root-1 MuiTypography-body1-10 MuiTypography-colorTextPrimary-23"
      style={
        Object {
          "color": "#E53935",
        }
      }
    >
      
    </h2>
    <button
      className="MuiButtonBase-root-88 MuiButton-root-62 MuiButton-text-64 MuiButton-outlined-70 MuiButton-sizeSmall-85"
      disabled={false}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "marginTop": 10,
        }
      }
      tabIndex="0"
      type="submit"
    >
      <span
        className="MuiButton-label-63"
      >
        <div
          className="MuiCircularProgress-root-98 MuiCircularProgress-colorPrimary-101 MuiCircularProgress-indeterminate-100"
          role="progressbar"
          style={
            Object {
              "height": 22,
              "width": 22,
            }
          }
        >
          <svg
            className="MuiCircularProgress-svg-103"
            viewBox="22 22 44 44"
          >
            <circle
              className="MuiCircularProgress-circle-104 MuiCircularProgress-circleIndeterminate-106"
              cx={44}
              cy={44}
              fill="none"
              r={20.2}
              strokeWidth={3.6}
              style={Object {}}
            />
          </svg>
        </div>
      </span>
      <span
        className="MuiTouchRipple-root-91"
      />
    </button>
  </form>
</div>
`;
